core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -582,6 +592,53 @@ private void maybeTerminateChannel() {
     }
   }
 
+  @Override
+  public ConnectivityState getState(boolean requestConnection) {
+    ConnectivityState savedChannelState = channelState;
+    if (requestConnection && savedChannelState == IDLE) {
+      channelExecutor.executeLater(new Runnable() {
+        @Override
+        public void run() {
+          exitIdleMode();
+        }
+      }).drain();
+    }
+    return savedChannelState;
+  }
+
+  private static final class StateChangeCallBack implements Runnable {
+    final ConnectivityState source;
+    final Runnable callback;
+    boolean done;
+
+    StateChangeCallBack(ConnectivityState source, Runnable callBack) {
+      this.source = source;
+      this.callback = callBack;
+    }
+
+    @Override
+    public synchronized void run() {
```

Comment:
yes, one possibility is the application thread calling `notifyWhenStateChanged`, the other is the transport thread calling `updateChannelState`. Anyway, this Queue implementation was abandoned in favor of ArrayList.